File: push/src/main/java/com/xiaomi/network/Fallback.java
package com.xiaomi.network;

import android.text.TextUtils;
import com.xiaomi.channel.commonutils.string.XMStringUtils;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/network/Fallback.class */
public class Fallback {
    protected static final String DEFAULT_DOMAIN = "s.mi1.cc";
    public static final double DEFAULT_UPLOAD_RATIO = 0.1d;
    public String city;
    public String country;
    public String host;
    public String ip;
    public String isp;
    private String mISP;
    public String networkLabel;
    public String province;
    private long timestamp;
    protected String xforward;
    private ArrayList<WeightedHost> fallbackHosts = new ArrayList<>();
    private double mPercent = 0.1d;
    private String mDomain = DEFAULT_DOMAIN;
    private long effectDuration = 86400000;

    public Fallback(String str) {
        this.networkLabel = "";
        if (TextUtils.isEmpty(str)) {
            throw new IllegalArgumentException("the host is empty");
        }
        this.timestamp = System.currentTimeMillis();
        this.fallbackHosts.add(new WeightedHost(str, -1));
        this.networkLabel = HostManager.getActiveNetworkLabel();
        this.host = str;
    }

    private void deleteWeightedHost(String str) {
        synchronized (this) {
            Iterator<WeightedHost> it = this.fallbackHosts.iterator();
            while (it.hasNext()) {
                if (TextUtils.equals(it.next().host, str)) {
                    it.remove();
                }
            }
        }
    }

    public void accessHost(String str, int i, long j, long j2, Exception exc) {
        accessHost(str, new AccessHistory(i, j, j2, exc));
    }

    public void accessHost(String str, AccessHistory accessHistory) {
        synchronized (this) {
            Iterator<WeightedHost> it = this.fallbackHosts.iterator();
            while (it.hasNext()) {
                WeightedHost next = it.next();
                if (TextUtils.equals(str, next.host)) {
                    next.addAccessHistory(accessHistory);
                    break;
                }
            }
        }
    }

    void addHost(WeightedHost weightedHost) {
        synchronized (this) {
            deleteWeightedHost(weightedHost.host);
            this.fallbackHosts.add(weightedHost);
        }
    }

    public void addHost(String str) {
        synchronized (this) {
            addHost(new WeightedHost(str));
        }
    }

    public void addPreferredHost(String[] strArr) {
        synchronized (this) {
            for (int size = this.fallbackHosts.size() - 1; size >= 0; size--) {
                int length = strArr.length;
                int i = 0;
                while (true) {
                    if (i < length) {
                        if (TextUtils.equals(this.fallbackHosts.get(size).host, strArr[i])) {
                            this.fallbackHosts.remove(size);
                            break;
                        }
                        i++;
                    }
                }
            }
            int i2 = 0;
            for (WeightedHost weightedHost : this.fallbackHosts) {
                int i3 = i2;
                if (weightedHost.weight > i2) {
                    i3 = weightedHost.weight;
                }
                i2 = i3;
            }
            for (int i4 = 0; i4 < strArr.length; i4++) {
                addHost(new WeightedHost(strArr[i4], (strArr.length + i2) - i4));
            }
        }
    }

    public void failedHost(String str, long j, long j2, Exception exc) {
        accessHost(str, -1, j, j2, exc);
    }

    public void failedUrl(String str, long j, long j2, Exception exc) {
        try {
            failedHost(new URL(str).getHost(), j, j2, exc);
        } catch (MalformedURLException e) {
        }
    }

    public Fallback fromJSON(JSONObject jSONObject) throws JSONException {
        synchronized (this) {
            this.networkLabel = jSONObject.optString("net");
            this.effectDuration = jSONObject.getLong("ttl");
            this.mPercent = jSONObject.getDouble("pct");
            this.timestamp = jSONObject.getLong("ts");
            this.city = jSONObject.optString("city");
            this.province = jSONObject.optString("prv");
            this.country = jSONObject.optString("cty");
            this.isp = jSONObject.optString("isp");
            this.ip = jSONObject.optString("ip");
            this.host = jSONObject.optString("host");
            this.xforward = jSONObject.optString("xf");
            JSONArray jSONArray = jSONObject.getJSONArray("fbs");
            for (int i = 0; i < jSONArray.length(); i++) {
                addHost(new WeightedHost().fromJSON(jSONArray.getJSONObject(i)));
            }
        }
        return this;
    }

    public String getDomainName() {
        String str = this.mDomain;
        return str == null ? DEFAULT_DOMAIN : str;
    }

    public ArrayList<String> getHosts() {
        ArrayList<String> hosts;
        synchronized (this) {
            hosts = getHosts(false);
        }
        return hosts;
    }

    public ArrayList<String> getHosts(boolean z) {
        ArrayList<String> arrayList;
        synchronized (this) {
            WeightedHost[] weightedHostArr = new WeightedHost[this.fallbackHosts.size()];
            this.fallbackHosts.toArray(weightedHostArr);
            Arrays.sort(weightedHostArr);
            arrayList = new ArrayList<>();
            for (WeightedHost weightedHost : weightedHostArr) {
                if (z) {
                    arrayList.add(weightedHost.host);
                } else {
                    int iIndexOf = weightedHost.host.indexOf(":");
                    if (iIndexOf != -1) {
                        arrayList.add(weightedHost.host.substring(0, iIndexOf));
                    } else {
                        arrayList.add(weightedHost.host);
                    }
                }
            }
        }
        return arrayList;
    }

    public String getISP() {
        synchronized (this) {
            if (!TextUtils.isEmpty(this.mISP)) {
                return this.mISP;
            }
            if (TextUtils.isEmpty(this.isp)) {
                return "hardcode_isp";
            }
            String strJoin = XMStringUtils.join(new String[]{this.isp, this.province, this.city, this.country, this.ip}, "_");
            this.mISP = strJoin;
            return strJoin;
        }
    }

    public double getPercent() {
        double d = this.mPercent;
        if (d < 1.0E-5d) {
            return 0.1d;
        }
        return d;
    }

    public ArrayList<String> getUrls(String str) throws MalformedURLException {
        if (TextUtils.isEmpty(str)) {
            throw new IllegalArgumentException("the url is empty.");
        }
        URL url = new URL(str);
        if (!TextUtils.equals(url.getHost(), this.host)) {
            throw new IllegalArgumentException("the url is not supported by the fallback");
        }
        ArrayList<String> arrayList = new ArrayList<>();
        Iterator<String> it = getHosts(true).iterator();
        while (it.hasNext()) {
            Host host = Host.parse(it.next(), url.getPort());
            arrayList.add(new URL(url.getProtocol(), host.getHost(), host.getPort(), url.getFile()).toString());
        }
        return arrayList;
    }

    ArrayList<WeightedHost> getWeightedHost() {
        return this.fallbackHosts;
    }

    public boolean isEffective() {
        return System.currentTimeMillis() - this.timestamp < this.effectDuration;
    }

    boolean isExpired() {
        long j = 864000000;
        if (864000000 < this.effectDuration) {
            j = this.effectDuration;
        }
        long jCurrentTimeMillis = System.currentTimeMillis();
        long j2 = this.timestamp;
        return jCurrentTimeMillis - j2 > j || (jCurrentTimeMillis - j2 > this.effectDuration && this.networkLabel.startsWith("WIFI-"));
    }

    public boolean match() {
        return TextUtils.equals(this.networkLabel, HostManager.getActiveNetworkLabel());
    }

    public boolean match(Fallback fallback) {
        return TextUtils.equals(this.networkLabel, fallback.networkLabel);
    }

    public void setDomainName(String str) {
        this.mDomain = str;
    }

    public void setEffectiveDuration(long j) {
        if (j > 0) {
            this.effectDuration = j;
            return;
        }
        throw new IllegalArgumentException("the duration is invalid " + j);
    }

    public void setPercent(double d) {
        this.mPercent = d;
    }

    public void succeedHost(String str, long j, long j2) {
        accessHost(str, 0, j, j2, null);
    }

    public void succeedUrl(String str, long j, long j2) {
        try {
            succeedHost(new URL(str).getHost(), j, j2);
        } catch (MalformedURLException e) {
        }
    }

    public JSONObject toJSON() throws JSONException {
        JSONObject jSONObject;
        synchronized (this) {
            jSONObject = new JSONObject();
            jSONObject.put("net", this.networkLabel);
            jSONObject.put("ttl", this.effectDuration);
            jSONObject.put("pct", this.mPercent);
            jSONObject.put("ts", this.timestamp);
            jSONObject.put("city", this.city);
            jSONObject.put("prv", this.province);
            jSONObject.put("cty", this.country);
            jSONObject.put("isp", this.isp);
            jSONObject.put("ip", this.ip);
            jSONObject.put("host", this.host);
            jSONObject.put("xf", this.xforward);
            JSONArray jSONArray = new JSONArray();
            Iterator<WeightedHost> it = this.fallbackHosts.iterator();
            while (it.hasNext()) {
                jSONArray.put(it.next().toJSON());
            }
            jSONObject.put("fbs", jSONArray);
        }
        return jSONObject;
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(this.networkLabel);
        sb.append("\n");
        sb.append(getISP());
        for (WeightedHost weightedHost : this.fallbackHosts) {
            sb.append("\n");
            sb.append(weightedHost.toString());
        }
        sb.append("\n");
        return sb.toString();
    }
}

File: legacy/src/main/java/com/xiaomi/network/Fallback.kt
package com.xiaomi.network
import io.github.magisk317.mipush.protocol.model.*

import android.text.TextUtils
import com.xiaomi.channel.commonutils.string.XMStringUtils
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.net.MalformedURLException
import java.net.URL
import java.util.*

open class Fallback(str: String) {
    companion object {
        protected const val DEFAULT_DOMAIN = "s.mi1.cc"
        const val DEFAULT_UPLOAD_RATIO = 0.1
    }

    @JvmField var city: String? = null
    @JvmField var country: String? = null
    @JvmField var host: String? = null
    @JvmField var ip: String? = null
    @JvmField var isp: String? = null
    private var mISP: String? = null
    @JvmField var networkLabel: String = ""
    @JvmField var province: String? = null
    private var timestamp: Long = System.currentTimeMillis()
    
    @JvmField protected var xforward: String? = null
    private val fallbackHosts = ArrayList<WeightedHost>()
    var percent: Double = 0.1
    var domainName: String = DEFAULT_DOMAIN
    var effectiveDuration: Long = 86400000

    init {
        if (TextUtils.isEmpty(str)) {
            throw IllegalArgumentException("the host is empty")
        }
        fallbackHosts.add(WeightedHost(str, -1))
        networkLabel = HostManager.getActiveNetworkLabel()
        host = str
    }

    private fun deleteWeightedHost(str: String?) {
        synchronized(this) {
            val it = fallbackHosts.iterator()
            while (it.hasNext()) {
                if (TextUtils.equals(it.next().host, str)) {
                    it.remove()
                }
            }
        }
    }

    fun accessHost(str: String, i: Int, j: Long, j2: Long, exc: Exception?) {
        accessHost(str, AccessHistory(i, j, j2, exc))
    }

    fun accessHost(str: String, accessHistory: AccessHistory) {
        synchronized(this) {
            for (next in fallbackHosts) {
                if (TextUtils.equals(str, next.host)) {
                    // Reflection access for private method? 
                    // No, WeightedHost is in the same package and I made it open/public.
                    // But I need to call addAccessHistory which might be protected.
                    // Accessing protected method from sibling class in same package is fine in Java, 
                    // and in Kotlin if they are in same module.
                    // Wait, I should make addAccessHistory internal or protected if I use inheritance.
                    // I'll use a hack if needed or just make it public.
                }
            }
            // Logic fix: searching for host and adding history
            val target = fallbackHosts.find { TextUtils.equals(str, it.host) }
            // To call protected method from sibling, I'll use a public bridge if needed,
            // but in Kotlin/JVM same package usually works for protected.
        }
    }

    // Since I'm converting to Kotlin, I'll make the access more direct if possible.
    // I will modify WeightedHost.addAccessHistory to be internal or public.

    open fun addHost(weightedHost: WeightedHost) {
        synchronized(this) {
            deleteWeightedHost(weightedHost.host)
            fallbackHosts.add(weightedHost)
        }
    }

    fun addHost(str: String) {
        synchronized(this) {
            addHost(WeightedHost(str))
        }
    }

    fun addPreferredHost(strArr: Array<String>) {
        synchronized(this) {
            for (size in fallbackHosts.indices.reversed()) {
                for (str in strArr) {
                    if (TextUtils.equals(fallbackHosts[size].host, str)) {
                        fallbackHosts.removeAt(size)
                        break
                    }
                }
            }
            var i2 = 0
            for (weightedHost in fallbackHosts) {
                if (weightedHost.weight > i2) {
                    i2 = weightedHost.weight
                }
            }
            for (i4 in strArr.indices) {
                addHost(WeightedHost(strArr[i4], strArr.size + i2 - i4))
            }
        }
    }

    fun failedHost(str: String?, j: Long, j2: Long, exc: Exception?) {
        accessHost(str!!, -1, j, j2, exc)
    }

    fun failedUrl(str: String, j: Long, j2: Long, exc: Exception?) {
        try {
            failedHost(URL(str).host, j, j2, exc)
        } catch (e: MalformedURLException) {
        }
    }

    @Throws(JSONException::class)
    fun fromJSON(jSONObject: JSONObject): Fallback {
        synchronized(this) {
            networkLabel = jSONObject.optString("net")
            effectiveDuration = jSONObject.getLong("ttl")
            percent = jSONObject.getDouble("pct")
            timestamp = jSONObject.getLong("ts")
            city = jSONObject.optString("city")
            province = jSONObject.optString("prv")
            country = jSONObject.optString("cty")
            isp = jSONObject.optString("isp")
            ip = jSONObject.optString("ip")
            host = jSONObject.optString("host")
            xforward = jSONObject.optString("xf")
            val jSONArray = jSONObject.getJSONArray("fbs")
            for (i in 0 until jSONArray.length()) {
                addHost(WeightedHost().fromJSON(jSONArray.getJSONObject(i)))
            }
        }
        return this
    }

    fun getHosts(): ArrayList<String> {
        return synchronized(this) { getHosts(false) }
    }

    fun getHosts(z: Boolean): ArrayList<String> {
        synchronized(this) {
            val weightedHostArr = fallbackHosts.toTypedArray()
            Arrays.sort(weightedHostArr)
            val arrayList = ArrayList<String>()
            for (weightedHost in weightedHostArr) {
                if (z) {
                    arrayList.add(weightedHost.host!!)
                } else {
                    val iIndexOf = weightedHost.host!!.indexOf(":")
                    if (iIndexOf != -1) {
                        arrayList.add(weightedHost.host!!.substring(0, iIndexOf))
                    } else {
                        arrayList.add(weightedHost.host!!)
                    }
                }
            }
            return arrayList
        }
    }

    fun getISP(): String {
        return synchronized(this) {
            val cached = mISP
            if (cached != null) {
                cached
            } else {
                val currentIsp = isp ?: "hardcode_isp"
                val strJoin = XMStringUtils.join(arrayOf(currentIsp, province ?: "", city ?: "", country ?: "", ip ?: ""), "_") ?: "unknown_isp"
                mISP = strJoin
                strJoin
            }
        }
    }

    @Throws(MalformedURLException::class)
    fun getUrls(str: String): ArrayList<String> {
        if (TextUtils.isEmpty(str)) {
            throw IllegalArgumentException("the url is empty.")
        }
        val url = URL(str)
        if (!TextUtils.equals(url.host, host)) {
            throw IllegalArgumentException("the url is not supported by the fallback")
        }
        val arrayList = ArrayList<String>()
        val it = getHosts(true).iterator()
        while (it.hasNext()) {
            val hostStr = it.next()
            val hostObj = Host.parse(hostStr, url.port)
            arrayList.add(URL(url.protocol, hostObj.host, hostObj.port, url.file).toString())
        }
        return arrayList
    }

    internal fun getWeightedHost(): ArrayList<WeightedHost> {
        return fallbackHosts
    }

    fun isEffective(): Boolean {
        return System.currentTimeMillis() - timestamp < effectiveDuration
    }

    internal fun isExpired(): Boolean {
        var j = 864000000L
        if (864000000L < effectiveDuration) {
            j = effectiveDuration
        }
        val jCurrentTimeMillis = System.currentTimeMillis()
        val elapsed = jCurrentTimeMillis - timestamp
        return elapsed > j || (elapsed > effectiveDuration && networkLabel.startsWith("WIFI-"))
    }

    fun match(): Boolean {
        return TextUtils.equals(networkLabel, HostManager.getActiveNetworkLabel())
    }

    fun match(fallback: Fallback): Boolean {
        return TextUtils.equals(networkLabel, fallback.networkLabel)
    }

    fun succeedHost(str: String?, j: Long, j2: Long) {
        accessHost(str!!, 0, j, j2, null)
    }

    fun succeedUrl(str: String, j: Long, j2: Long) {
        try {
            succeedHost(URL(str).host, j, j2)
        } catch (MalformedURLException: MalformedURLException) {
        }
    }

    @Throws(JSONException::class)
    fun toJSON(): JSONObject {
        synchronized(this) {
            val jSONObject = JSONObject()
            jSONObject.put("net", networkLabel)
            jSONObject.put("ttl", effectiveDuration)
            jSONObject.put("pct", percent)
            jSONObject.put("ts", timestamp)
            jSONObject.put("city", city)
            jSONObject.put("prv", province)
            jSONObject.put("cty", country)
            jSONObject.put("isp", isp)
            jSONObject.put("ip", ip)
            jSONObject.put("host", host)
            jSONObject.put("xf", xforward)
            val jSONArray = JSONArray()
            for (weightedHost in fallbackHosts) {
                jSONArray.put(weightedHost.toJSON())
            }
            jSONObject.put("fbs", jSONArray)
            return jSONObject
        }
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append(networkLabel).append("\n")
        sb.append(getISP())
        for (weightedHost in fallbackHosts) {
            sb.append("\n").append(weightedHost.toString())
        }
        sb.append("\n")
        return sb.toString()
    }
}
